### DISCUSSION TITLE
Grenzwerte mit zyklischen Telegrammen

### DISCUSSION THREAD

**Discussion by chef76** [brutella/hkknx-public#109] [Q&A] [answered]
Hallo Matthias,

Ich bräuchte mal kurz Deine Unterstützung. Das nachfolgende Skript schließt meine Lamellen der Jalousien, wenn die Raumtemperatur über 23,5 °C geht.

```go
var hkknx = import("hkknx")
temp = hkknx.GroupWriteNotify("2/1/28")

for {
    select {
    case buf = <- temp:
        if hkknx.ParseDPT9001(buf) > 23.5 {  
            openCloseAddrs = []string{"1/1/8", "1/1/10", "1/1/12", "1/1/14", "1/1/16", "1/2/4", "1/2/12"}
            hkknx.GroupWrite(true, openCloseAddrs...)
        } else {
            angleAddrs = []string{"1/1/45","1/1/37", "1/1/61", "1/1/65", "1/1/49", "1/2/35", "1/2/51"}
            hkknx.GroupWrite(hkknx.DPT5001(40), angleAddrs...)
        }
   }
}
```
Da aber der Temperaturwert zyklisch gesendet wird (bei Änderung oder alle Minute), werden die Aktoren mit jedem Telegramm angesteuert. Kannst Du mir das Skript so umbauen, dass nur bei Über-, bzw. Unterschreiten der Grenze einmal der jeweilige Schleifenteil ausgeführt wird?

Ich hab in allen verfügbaren Skripten leider nichts gefunden, was ich dafür hätte umbauen können. :-(

Vielen Dank im Voraus!

**Comment by brutella** [comment, marked as answer]
Um das zu lösen, musst du in einer Variable speichern, was das letzte Mal ausgeführt wurde.
Ich würde daher die Vergleich-Operation in einer Variable speichern.

```go
var shouldClose = hkknx.ParseDPT9001(buf) > 23.5
```

Somit weisst du schonmal, ob die Lamellen geschlossen oder geöffnet werden sollen.
Diese Variable verwendest du dann in dem If-Statement.

```go
if shouldClose {
    // schließen
} else {
    // öffnen
}
```
Nach dem Öffnen und schließen, speicherst du in einer Variable, den Wert von `shouldClose`.

```go
isClosed = shouldClose
```

Die Variable `isClosed` definierst du aber außerhalb der Schleife, damit der Wert beim nächsten Schleifendurchlauf erhalten bleibt.
Da die Variable keinen Anfangswert haben soll (du weißt ja nicht, ob die Lamelle beim Start des Skripts geöffnet oder geschlossen ist), hat sie den Wert `nil` - also nichts.

Somit kannst du die Variablen `isClosed` und `shouldClose` miteinander vergleichen. 
Sind beide Werte gleich, dann musst du nichts auf den Bus senden.
Sind die Werte ungleich, muss man etwas auf den Bus senden.

Das ganze Skript schaut folgendermaßen aus.

```go
var hkknx = import("hkknx")
var temp = hkknx.GroupWriteNotify("2/1/28")
var isClosed = nil

for {
   buf = <-temp

   // shouldClose ist entweder *true* oder *false*, je nachdem ob die Temperature größer oder kleiner als 23.5 ist.
   var shouldClose = hkknx.ParseDPT9001(buf) > 23.5

   // Überprüfe ob shouldClose gleich isClosed ist.
   if shouldClose == isClosed {
       continue
   }

   if shouldClose {
       openCloseAddrs = []string{"1/1/8", "1/1/10", "1/1/12", "1/1/14", "1/1/16", "1/2/4", "1/2/12"}
       err = hkknx.GroupWrite(true, openCloseAddrs...)
       if err != nil {
           println(err)
           continue
       }
   } else {
       angleAddrs = []string{"1/1/45","1/1/37", "1/1/61", "1/1/65", "1/1/49", "1/2/35", "1/2/51"}
       err = hkknx.GroupWrite(hkknx.DPT5001(40), angleAddrs...)
       if err != nil {
           println(err)
           continue
       }
   }

   isClosed = shouldClose
}
```

Die entscheidende Überprüfung ist 

```go
if shouldClose == isClosed {
    continue
}
```

Beim Statement *continue* springt man wieder an den Anfang der For-Schleife. 
Alles darunter wird nicht ausgeführt.

**Reply by chef76** [reply]
Hallo Matthias,

verstanden Danke! einen kleinen Test hast Du auch noch eingebaut isClose**d** = shouldClose. :-)

**Reply by brutella** [reply]
Ist behoben. 😉
